### PR TITLE
ARROW-15745: [Java] Deprecate redundant iterable of ScanTask

### PR DESCRIPTION
**Pull request by davisusanibar**
Deprecate redundant iterable ScanTask since there are no more ScanTasks on the C++ side

### CODE REVIEW

File: java/dataset/src/main/java/org/apache/arrow/dataset/jni/NativeScanner.java [hunk 1]
```diff
@@ -68,6 +68,15 @@ ArrowReader execute() {
   }
 
   @Override
+  public NativeScanTask scanTask() {
```
Comment (lidavidm) [inline java/dataset/src/main/java/org/apache/arrow/dataset/jni/NativeScanner.java:71]: We just shouldn't have ScanTask at all, there is no such thing on the C++ side anymore.

**Review by lidavidm — CHANGES_REQUESTED** [review]

File: java/dataset/src/main/java/org/apache/arrow/dataset/jni/NativeScanner.java (hunk 1 above)
Comment (davisusanibar) [reply java/dataset/src/main/java/org/apache/arrow/dataset/jni/NativeScanner.java:71]: Ok, let me delete that, this `scan()` method is exposed to the client and will cause break compatibilities

Comment (lidavidm) [reply java/dataset/src/main/java/org/apache/arrow/dataset/jni/NativeScanner.java:71]: The deprecation makes sense to me, I think we should consider what the new interface is, though. I would think something like `ArrowReader scanBatchesUnordered()`.

Comment (lidavidm) [reply java/dataset/src/main/java/org/apache/arrow/dataset/jni/NativeScanner.java:71]: That would mirror the C++ API

Comment (davisusanibar) [reply java/dataset/src/main/java/org/apache/arrow/dataset/jni/NativeScanner.java]: Changed

File: java/dataset/src/main/java/org/apache/arrow/dataset/scanner/Scanner.java [hunk 2]
```diff
@@ -17,19 +17,29 @@
...
+import org.apache.arrow.vector.ipc.ArrowReader;
 import org.apache.arrow.vector.types.pojo.Schema;
 
 /**
  * A high level interface for scanning data over dataset.
  */
 public interface Scanner extends AutoCloseable {
 
+  /**
+   * Read record batches from the URI resources filtering the batch size configured.
```
Comment (lidavidm) [inline java/dataset/src/main/java/org/apache/arrow/dataset/scanner/Scanner.java:29]:
```suggestion
   * Read the dataset as a stream of record batches.
```

File: java/dataset/src/main/java/org/apache/arrow/dataset/scanner/Scanner.java [hunk 3]
```diff
@@ -17,19 +17,29 @@
...
+   *
+   * @return a {@link ArrowReader}.
+   */
+  ArrowReader scanBatchesUnordered();
```
Comment (lidavidm) [inline java/dataset/src/main/java/org/apache/arrow/dataset/scanner/Scanner.java:33]: Sorry, it looks like it's actually `scanBatches` in the implementation: https://github.com/apache/arrow/blob/2629f208cf8be638c7e251946947009bc181bb31/java/dataset/src/main/cpp/jni_wrapper.cc#L152

File: java/dataset/src/main/java/org/apache/arrow/dataset/jni/NativeScanner.java [hunk 4]
```diff
@@ -68,6 +68,19 @@ ArrowReader execute() {
   }
 
   @Override
+  public ArrowReader scanBatchesUnordered() {
+    if (closed) {
+      throw new NativeInstanceReleasedException();
+    }
+    if (!executed.compareAndSet(false, true)) {
+      throw new UnsupportedOperationException("NativeScanner cannot be executed more than once. Consider creating " +
+              "new scanner instead");
```
Comment (lidavidm) [inline java/dataset/src/main/java/org/apache/arrow/dataset/jni/NativeScanner.java:77]:
```suggestion
      throw new UnsupportedOperationException("NativeScanner can only be executed once. Create a " +
              "new scanner instead");
```

File: java/dataset/src/main/java/org/apache/arrow/dataset/scanner/Scanner.java (hunk 3 above)
Comment (davisusanibar) [reply java/dataset/src/main/java/org/apache/arrow/dataset/scanner/Scanner.java]: Thanks, updated

**Review by lidavidm — APPROVED** [review]

**Pull request merged**